src/v/cluster/tm_stm.h

```diff
@@ -42,6 +43,64 @@ namespace cluster {
 using use_tx_version_with_last_pid_bool
   = ss::bool_class<struct use_tx_version_with_last_pid_tag>;
 
+class tm_stm;
+
+class txlock_unit {
```

Comment:
are there guarantees that this class will not outlive the `tm_stm` ? it is holding raw pointer to it

Reply:
tm_stm holds the mutex so if unit outlives tm_stm it means that it also outlives its mutex which shouldn't happen. Seastar uses raw pointer in its unit impl. so this change is at least as bad as as ss::semaphore